src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -111,5 +119,166 @@ class EntityClient (requestContext: RequestContext) extends Actor {
+      case Success(memberTypeOpt) => op(memberTypeOpt)
+    }
+  }
+
+  //Returns the plural form of the entity type.
+  //Bails with a 400 Bad Request if the entity type is unknown to the schema.
+  private def withPlural(entityType: String)(op: (String => Unit)): Unit = {
+    ModelSchema.getPlural(entityType) match {
+      case Failure(regret) => requestContext.complete(HttpResponse(BadRequest, regret.getMessage))
+      case Success(plural) => op(plural)
+    }
+  }
+
+  //Bail with a 400 Bad Request if the first column of the tsv has duplicate values.
+  //Otherwise, carry on.
+  private def checkFirstColumnsDistinct( tsv: TSVLoadFile )(op: => Unit): Unit = {
+    val entitiesToUpdate = tsv.tsvData.map(_(0))
+    val distinctEntities = entitiesToUpdate.distinct
+    if ( entitiesToUpdate.size != distinctEntities.size ) {
+      requestContext.complete( HttpResponse(BadRequest, "TSV has duplicated entities: " + entitiesToUpdate.diff(distinctEntities).distinct.mkString(", ")) )
+    } else {
+      op
+    }
+  }
+
+  //Bail with a 400 Bad Request if the tsv is trying to set members on a collection type.
+  //Otherwise, carry on.
+  private def checkNoCollectionMemberAttribute( tsv: TSVLoadFile, memberTypeOpt: Option[String] )(op: => Unit): Unit = {
+    if( memberTypeOpt.isDefined && tsv.headers.contains(memberTypeOpt.get + "_id") ) {
+      requestContext.complete( HttpResponse(BadRequest, "Can't set collection members along with other attributes; please use two-column TSV format or remove " + memberTypeOpt.get + "_id from your tsv.") )
+    } else {
+      op
+    }
+  }
+
+  //Verifies that the provided list of headers includes all attributes required by the schema for this entity type.
+  //Bails with a 400 Bad Request if the entity type is unknown or if some attributes are missing.
+  //Returns the list of required attributes if all is well.
+  private def withRequiredAttributes(entityType: String, headers: Seq[String])(op: (Map[String, String] => Unit)):Unit = {
+    ModelSchema.getRequiredAttributes(entityType) match {
+      case Failure(regret) => requestContext.complete(HttpResponse(BadRequest, regret.getMessage))
+      case Success(requiredAttributes) => {
+        if( !requiredAttributes.keySet.subsetOf(headers.toSet) ) {
+          requestContext.complete( HttpResponse(BadRequest, "TSV is missing required attributes: " + (requiredAttributes.keySet -- headers).mkString(", ")) )
+        } else {
+          op(requiredAttributes)
+        }
+      }
+    }
+  }
+
+  def batchCallToRawls( workspaceNamespace: String, workspaceName: String, calls: Seq[EntityUpdateDefinition] ) = {
+    log.info("TSV upload request received")
+    val pipeline: HttpRequest => Future[HttpResponse] =
+      addHeader(Cookie(requestContext.request.cookies)) ~> sendReceive
+    val responseFuture: Future[HttpResponse] = pipeline {
+      Post(FireCloudConfig.Workspace.entityPathFromWorkspace(workspaceNamespace, workspaceName)+"/batchUpsert",
```

Comment:
In general, we've tried to keep all string components of url paths (i.e. `"/batchUpsert"`) in the configuration instead of in code. However, we are going to look into refactoring that in the near future - it's probably not worth doing here. Others may disagree.